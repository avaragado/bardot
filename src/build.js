// @flow

import chalk from 'chalk';

type WidthFill = {| mode: 'fill', ctCharMinus: number, ctCharFull: ?number |};
type WidthBar = {| mode: 'bar', ctChar: number |};
type WidthTemplate = {| mode: 'template', ctChar: number |};

type Template = string;

type Width = WidthFill | WidthBar | WidthTemplate;

type BarSymbol = {|
    full: string,
    empty: string,
    fractions: string[],
|};

export type Option = {
    cur: number,
    max: number,
    width: Width,
    tpl: Template,
    symbol: BarSymbol,
    renderer: (opt: Option) => string,
};

type OptionSparse = {
    cur?: number,
    max?: number,
    width?: Width,
    tpl?: Template,
    symbol?: BarSymbol,
    renderer?: (opt: Option) => string,
};

type Chainable = {
    current: (cur: number) => Chainable,
    maximum: (max: number) => Chainable,
    widthFill: (ctCharMinus?: number, ctCharFull?: ?number) => Chainable,
    widthBar: (ctChar?: number) => Chainable,
    widthTemplate: (ctChar?: number) => Chainable,
    template: (tpl: Template) => Chainable,
    symbols: (symbol: BarSymbol) => Chainable,
    toString: () => string,
    opt: () => Option,
};

export const width: { [key: string]: (ct?: number) => Width } = {
    fill: (ctCharMinus = 0, ctCharFull = null) => ({ mode: 'fill', ctCharMinus, ctCharFull }),
    bar: (ctChar = 50) => ({ mode: 'bar', ctChar }),
    template: (ctChar = 50) => ({ mode: 'template', ctChar }),
};

const greenbar = chalk.green('|bar|');
const dimslash = chalk.dim('/');

export const template: { [key: string]: Template } = {
    bar: `${greenbar}`,
    barCur: `${greenbar} |cur|`,
    barCurMax: `${greenbar} |cur|${dimslash}|max|`,
    barPct: `${greenbar} |pct|%`,
    barCurMaxPct: `${greenbar} |cur|${dimslash}|max| |pct|%`,
};

export const symbol: { [key: string]: BarSymbol } = {
    dot8: {
        full: '⣿',
        empty: ' ',
        fractions: ['⡀', '⡄', '⡆', '⡇', '⣇', '⣧', '⣷'],
    },
    dot6: {
        full: '⠿',
        empty: ' ',
        fractions: ['⠄', '⠆', '⠇', '⠧', '⠷'],
    },
    rod5: {
        full: '𝍤',
        empty: ' ',
        fractions: ['𝍠', '𝍡', '𝍢', '𝍣'],
    },
    pip: {
        full: '●',
        empty: '○',
        fractions: [],
    },
    blockspace: {
        full: '█',
        empty: ' ',
        fractions: [],
    },
    blockdot: {
        full: '█',
        empty: '⠿',
        fractions: [],
    },
    tick: {
        full: '✓',
        empty: ' ',
        fractions: [],
    },
    starspace: {
        full: '*',
        empty: ' ',
        fractions: [],
    },
    hashdash: {
        full: '#',
        empty: '-',
        fractions: ['+', '⧺'],
    },
};

export const optDefault: Option = {
    cur: 0,
    max: 100,
    width: width.fill(0),
    tpl: template.barCurMax,
    symbol: symbol.dot8,
    renderer: () => '',
};

const build = (optIn?: OptionSparse): Chainable => {
    const opt = { ...optDefault, ...optIn };
    const extend = (optPlus: OptionSparse) => build({ ...opt, ...optPlus });
    const roundPos = val => Math.max(0, Math.round(val));

    return {
        current: cur => extend({ cur: Math.min(opt.max, roundPos(cur)) }),
        maximum: max => extend({ max: roundPos(max), cur: Math.min(roundPos(max), opt.cur) }),
        widthFill: (ctCharMinus, ctCharFull) =>
            extend({ width: width.fill(ctCharMinus, ctCharFull) }),
        widthBar: ctChar => extend({ width: width.bar(ctChar) }),
        widthTemplate: ctChar => extend({ width: width.template(ctChar) }),
        template: tpl => extend({ tpl }),
        symbols: symbolIn => extend({ symbol: symbolIn }),
        toString: () => opt.renderer(opt),
        opt: () => opt,
    };
};

export default build;
