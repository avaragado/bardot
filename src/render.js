// @flow

import len from 'string-length'; // ignores ansi escape sequences

import type { Option } from './build';

type OptData = {
    opt: Option,
    data: {
        ctCharBar: number,
        ctBlockPerNum: number,
        ctPipPerNum: number,
    },
};

type ReplaceMap = {
    [key: string]: string,
};

const replace = (tpl: string, map: ReplaceMap): string => tpl.replace(
    /\|(bar|cur|max|pct)\|/g,
    (match: string, word: string): string => map[word],
);

const ctCharLabelMax = (opt: Option): number => len(replace(opt.tpl, {
    cur: opt.max.toString(),
    max: opt.max.toString(),
    pct: '99.9', // more chars than '100'
    bar: '',
}));

const ctCharBarDerive = (width, opt) => {
    switch (width.mode) {
        // simplest case: the size of the bar is exactly as configured
        case 'bar': {
            return width.ctChar;
        }

        // user configures the total size of bar + label.
        // derive the bar size by subtracting the maximum label size.
        case 'template': {
            return width.ctChar - ctCharLabelMax(opt);
        }

        // user wants to fill the available width, minus the configured number of characters.
        case 'fill': {
            // $FlowFixMe flow moans at process.stdout.columns, no idea how to fix
            const ctCharFull = width.ctCharFull || process.stdout.columns;

            return ctCharFull - width.ctCharMinus - ctCharLabelMax(opt);
        }

        default: {
            return 0;
        }
    }
};

const addData = (opt: Option): OptData => {
    const ctCharBar = ctCharBarDerive(opt.width, opt);

    return {
        opt,
        data: {
            ctCharBar,
            ctBlockPerNum: ctCharBar / opt.max,
            ctPipPerNum: (ctCharBar * (opt.symbol.fractions.length + 1)) / opt.max,
        },
    };
};

const bar = ({ opt, data }) => {
    const ctBlock = opt.cur * data.ctBlockPerNum;
    const sBlocks = opt.symbol.full.repeat(ctBlock);
    const sEmpties = opt.symbol.empty.repeat((opt.max - opt.cur) * data.ctBlockPerNum);

    if (ctBlock === Math.floor(ctBlock)) {
        return sBlocks + sEmpties;
    }

    const arsPart = [opt.symbol.empty].concat(opt.symbol.fractions || []);
    const sPart = arsPart[Math.floor(opt.cur * data.ctPipPerNum) % arsPart.length];

    return sBlocks + sPart + sEmpties;
};

const renderOptData = (od) => {
    if (od.data.ctCharBar < 0) {
        return ''; // no room: return empty string rather than junk
    }

    const map = {
        // right-align cur and pct within their space to avoid overall length changes
        cur: od.opt.cur.toString().padStart(od.opt.max.toString().length),
        max: od.opt.max.toString(),
        pct: (Math.round((1000 * od.opt.cur) / od.opt.max) / 10).toString().padStart(4),
        bar: bar(od),
    };

    return replace(od.opt.tpl, map);
};

export default (opt: Option): string => renderOptData(addData(opt));
