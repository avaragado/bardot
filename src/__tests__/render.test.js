// @flow

import chalk from 'chalk';

import type { Option } from '../build';
import { width, symbol } from '../build';
import render from '../render';

const arOpt: Option[] = [
    {
        cur: 32,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 32,
        max: 32,
        width: width.bar(8),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 18,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 16,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 8,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 0,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 32,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max| |pct|%',
    },
    {
        cur: 18,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max| |pct|%',
    },
    {
        cur: 18,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max| |pct|% //// |bar| |cur|/|max| |pct|%',
    },
    {
        cur: 18,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: 'look |pct|% at |cur|/|max| this |bar|!',
    },
    {
        cur: 18,
        max: 32,
        width: width.template(10),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 18,
        max: 32,
        width: width.template(16),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max| |pct|%',
    },
    {
        cur: 18,
        max: 32,
        width: width.template(20),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max| |pct|%',
    },
    {
        cur: 18,
        max: 32,
        width: width.fill(10, 30),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max| |pct|%',
    },
    {
        cur: 10,
        max: 100,
        width: width.fill(0, 30),
        symbol: symbol.dot8,
        tpl: '%bar',
    },
    {
        cur: 10,
        max: 100,
        width: width.fill(0, 30),
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 100,
        max: 100,
        width: width.fill(0),
        symbol: symbol.dot8,
        tpl: `${chalk.green('|bar|')} |cur|/|max|`,
    },
    {
        cur: 18,
        max: 32,
        width: width.bar(4),
        symbol: symbol.dot8,
        tpl: [
            chalk.white.bgMagenta('|bar|'),
            ' ',
            chalk.bold('|cur|'),
            chalk.dim('/'),
            chalk.bold('|max|'),
            ' ',
            chalk.green('|pct|%'),
        ].join(''),
    },
    {
        cur: 18,
        max: 32,
        width: width.template(5), // not long enough
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 18,
        max: 32,
        width: width.fill(0, 5), // not long enough
        symbol: symbol.dot8,
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 18,
        max: 32,
        width: width.bar(4),
        symbol: {
            full: '*',
            empty: '-',
            fractions: [],
        },
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 20,
        max: 32,
        width: width.bar(4),
        symbol: {
            full: '#',
            empty: ' ',
            fractions: ['+'],
        },
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 20,
        max: 32,
        width: width.bar(4),
        symbol: {
            full: '[]',
            empty: '  ',
            fractions: ['..', '.+', '+.'],
        },
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 19,
        max: 32,
        width: width.bar(4),
        symbol: {
            full: 'x',
            empty: ' ',
            fractions: [],
        },
        tpl: '|bar| |cur|/|max|',
    },
    {
        cur: 19,
        max: 32,
        width: width.bar(4),
        symbol: {
            full: 'x',
            empty: ' ',
            fractions: ['o'],
        },
        tpl: '|bar| |cur|/|max|',
    },

].map(opt => ({ ...opt, renderer: () => '' })); // renderer ignored here

describe('render', () => {
    let col;
    const name = opt => `${opt.cur}/${opt.max} ${JSON.stringify(opt.width)} ${JSON.stringify(opt.symbol)} "${opt.tpl}"`;

    beforeEach(() => {
        // $FlowFixMe flow moans at process.stdout.columns, no idea how to fix
        col = process.stdout.columns;
        // $FlowFixMe flow moans at process.stdout.columns, no idea how to fix
        process.stdout.columns = 100;
    });

    afterEach(() => {
        // $FlowFixMe flow moans at process.stdout.columns, no idea how to fix
        process.stdout.columns = col;
    });

    arOpt.forEach((opt) => {
        it(name(opt), () => {
            const received = render(opt);
            // console.log(`<${received.replace(/ /g, 'â–¡')}> ${name(opt)}`);
            expect(received).toMatchSnapshot();
        });
    });
});
