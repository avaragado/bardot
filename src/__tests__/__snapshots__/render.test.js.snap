// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 0/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"      0/32"`;

exports[`render 8/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿     8/32"`;

exports[`render 10/100 {"mode":"fill","ctCharMinus":0,"ctCharFull":30} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "%bar" 1`] = `"%bar"`;

exports[`render 10/100 {"mode":"fill","ctCharMinus":0,"ctCharFull":30} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿â£¿â¡€                     10/100"`;

exports[`render 16/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿â£¿   16/32"`;

exports[`render 18/32 {"mode":"bar","ctChar":4} {"full":"*","empty":"-","fractions":[]} "|bar| |cur|/|max|" 1`] = `"**-- 18/32"`;

exports[`render 18/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "[37m[45m|bar|[49m[39m [1m|cur|[22m[2m/[22m[1m|max|[22m [32m|pct|%[39m" 1`] = `"[37m[45mâ£¿â£¿â¡„ [49m[39m [1m18[22m[2m/[22m[1m32[22m [32m56.3%[39m"`;

exports[`render 18/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max| |pct|% //// |bar| |cur|/|max| |pct|%" 1`] = `"â£¿â£¿â¡„  18/32 56.3% //// â£¿â£¿â¡„  18/32 56.3%"`;

exports[`render 18/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max| |pct|%" 1`] = `"â£¿â£¿â¡„  18/32 56.3%"`;

exports[`render 18/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿â£¿â¡„  18/32"`;

exports[`render 18/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "look |pct|% at |cur|/|max| this |bar|!" 1`] = `"look 56.3% at 18/32 this â£¿â£¿â¡„ !"`;

exports[`render 18/32 {"mode":"fill","ctCharMinus":0,"ctCharFull":5} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `""`;

exports[`render 18/32 {"mode":"fill","ctCharMinus":10,"ctCharFull":30} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max| |pct|%" 1`] = `"â£¿â£¿â£¿â£¿â¡‡    18/32 56.3%"`;

exports[`render 18/32 {"mode":"template","ctChar":5} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `""`;

exports[`render 18/32 {"mode":"template","ctChar":10} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿â£¿â¡„  18/32"`;

exports[`render 18/32 {"mode":"template","ctChar":16} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max| |pct|%" 1`] = `"â£¿â£¿â¡„  18/32 56.3%"`;

exports[`render 18/32 {"mode":"template","ctChar":20} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max| |pct|%" 1`] = `"â£¿â£¿â£¿â£¿â¡‡    18/32 56.3%"`;

exports[`render 19/32 {"mode":"bar","ctChar":4} {"full":"x","empty":" ","fractions":["o"]} "|bar| |cur|/|max|" 1`] = `"xx   19/32"`;

exports[`render 19/32 {"mode":"bar","ctChar":4} {"full":"x","empty":" ","fractions":[]} "|bar| |cur|/|max|" 1`] = `"xx   19/32"`;

exports[`render 20/32 {"mode":"bar","ctChar":4} {"full":"#","empty":" ","fractions":["+"]} "|bar| |cur|/|max|" 1`] = `"##+  20/32"`;

exports[`render 20/32 {"mode":"bar","ctChar":4} {"full":"[]","empty":"  ","fractions":["..",".+","+."]} "|bar| |cur|/|max|" 1`] = `"[][].+   20/32"`;

exports[`render 32/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max| |pct|%" 1`] = `"â£¿â£¿â£¿â£¿ 32/32  100%"`;

exports[`render 32/32 {"mode":"bar","ctChar":4} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿â£¿â£¿â£¿ 32/32"`;

exports[`render 32/32 {"mode":"bar","ctChar":8} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "|bar| |cur|/|max|" 1`] = `"â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿ 32/32"`;

exports[`render 100/100 {"mode":"fill","ctCharMinus":0,"ctCharFull":null} {"full":"â£¿","empty":" ","fractions":["â¡€","â¡„","â¡†","â¡‡","â£‡","â£§","â£·"]} "[32m|bar|[39m |cur|/|max|" 1`] = `"[32mâ£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿â£¿[39m 100/100"`;
